dev/ToastNotifications/ToastNotificationManager.cpp

```diff
@@ -140,10 +140,58 @@ namespace winrt::Microsoft::Windows::ToastNotifications::implementation
 
         return static_cast<winrt::Microsoft::Windows::ToastNotifications::ToastNotificationSetting>(toastNotificationSetting);
     }
-    winrt::Microsoft::Windows::ToastNotifications::ToastNotificationHistory ToastNotificationManager::History()
+    winrt::Windows::Foundation::IAsyncAction ToastNotificationManager::RemoveWithIdentiferAsync(uint32_t /*toastIdentifier*/)
     {
         throw hresult_not_implemented();
     }
+    winrt::Windows::Foundation::IAsyncAction ToastNotificationManager::RemoveWithTagAsync(hstring /*tag*/)
+    {
+        throw hresult_not_implemented();
+    }
+    winrt::Windows::Foundation::IAsyncAction ToastNotificationManager::RemoveWithTagGroupAsync(hstring /*tag*/, hstring /*group*/)
+    {
+        throw hresult_not_implemented();
+    }
+    winrt::Windows::Foundation::IAsyncAction ToastNotificationManager::RemoveGroupAsync(hstring /*group*/)
+    {
+        throw hresult_not_implemented();
+    }
+    winrt::Windows::Foundation::IAsyncAction ToastNotificationManager::RemoveAllAsync()
+    {
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveToastAppId() };
+        THROW_IF_FAILED(ToastNotifications_RemoveAllToastsForApp(appId.c_str()));
+
+        co_return;
+    }
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::ToastNotifications::ToastNotification>> ToastNotificationManager::GetAllAsync()
+    {
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveToastAppId() };
+        ABI::Windows::Foundation::Collections::IVector<ABI::Microsoft::Internal::ToastNotifications::INotificationProperties*>* toastPropertiesCollection;
+        THROW_IF_FAILED(ToastNotifications_GetHistory(appId.c_str(), &toastPropertiesCollection));
+
+        unsigned int count{};
+        THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+        
+        THROW_HR_IF(E_NOT_SET, count == 0);
```

Comment:
I don't think we should throw if the toast collection size is 0. This can be a valid scenario after all. What are your thoughts on this?

Reply:
I don't like to throw for things that aren't really errors either. That's the way the UDK does it, but that doesn't mean that's the way the WinAppSDK should do it. I could swallow the error and return an empty vector.

It also feels weird to create and empty vector to signal that there aren't any active notifications. I whish we could return an optional instead.

Reply:
Are you saying the UDK throws if the collection size is 0?

Reply:
Daniel is right. If there are no toasts in history, we should return an empty vector. Never throw unless there is a fatal error. So we should remove this check.

I hope we have a unit test that ensures that we don't throw and get an empty collection for this case.

Reply:
The WinAppSDK is now swallowing the error (E_NOT_SET) coming from CoreOS (thru the UDK) and returning an empty vector instead.

We have unit tests to verify the GetAllAsync logic with zero, one and many active toasts.

/rant -Building and returning and empty vector that the caller will have to destroy just to signal that there isn't anything to return, isxn't that much better than throwing an exception. I wish winrt had some support for more modern constructs, like optional.